src/warpychain/initial_syncer.go

```diff
@@ -0,0 +1,258 @@
+			if err != nil {
+				self.Log.WithError(err).Warn("Could not get last synced height")
+				return err
+			}
+
+			currentBlockHeight, currentBlockHash, err = self.warpyChainClient.getCurrentBlockHeight()
+			if err != nil {
+				self.Log.WithError(err).Warn("Could not get current block height")
+				return err
+			}
+
+			return err
+		})
+
+	if err != nil {
+		self.Log.WithError(err).Error("Could not get last synced block info")
+		// it requires an alert
+		self.monitor.GetReport().WarpyChainer.Errors.InitialSyncerLastSyncedError.Inc()
+		return
+	}
+
+	self.Log.WithField("lastSyncedHeight", lastSyncedHeight).WithField("currentBlockHeight", currentBlockHeight).
+		WithField("blocksToSync", currentBlockHeight-lastSyncedHeight).Info("Number of blocks to sync")
+
+	for ok := true; ok; ok = lastSyncedHeight < currentBlockHeight-1 {
+		blocksBatchSize := 20
+		if currentBlockHeight-lastSyncedHeight < 20 {
+			blocksBatchSize = int(currentBlockHeight - lastSyncedHeight)
+		}
+
+		blocksBatch := make([]int64, blocksBatchSize)
+		for i := range blocksBatch {
+			blocksBatch[i] = int64(i) + lastSyncedHeight + 1
+		}
+		lastSyncedHeight = self.syncBlocksInBatch(blocksBatch, ctx)
+
+		// Retrying infinitely?
+		err = task.NewRetry().
+			WithContext(self.Ctx).
+			WithMaxElapsedTime(0).
+			WithMaxInterval(self.maxInterval).
+			WithAcceptableDuration(self.maxInterval * 2).
+			WithOnError(func(err error, isDurationAcceptable bool) error {
+				if errors.Is(err, context.Canceled) && self.IsStopping.Load() {
+					return backoff.Permanent(err)
+				}
+
+				self.Log.WithError(err).Warn("Could not get current block info, retrying...")
+				return err
+			}).
+			Run(func() error {
+				currentBlockHeight, currentBlockHash, err = self.warpyChainClient.getCurrentBlockHeight()
+
+				return err
+			})
+	}
+
+	self.Log.WithField("currentBlockHeight", currentBlockHeight).WithField("currentBlockHash", currentBlockHash).
+		Info("No more blocks to sync, inserting last processed block height into sync_state")
+
+	// Retrying infinitely?
+	err = task.NewRetry().
+		WithContext(self.Ctx).
+		WithMaxElapsedTime(0).
+		WithMaxInterval(self.maxInterval).
+		WithAcceptableDuration(self.maxInterval * 2).
+		WithOnError(func(err error, isDurationAcceptable bool) error {
+			if errors.Is(err, context.Canceled) && self.IsStopping.Load() {
+				return backoff.Permanent(err)
+			}
+
+			self.Log.WithError(err).Warn("Could not update sync state, retrying...")
+			return err
+		}).
+		Run(func() error {
+			err = self.warpyChainClient.updateSyncState(ctx, currentBlockHeight, currentBlockHash)
+
+			return err
+		})
+
+	self.Log.WithField("currentBlockHeight", currentBlockHeight).WithField("currentBlockHash", currentBlockHash).
+		Info("sync_state updated with last processed block height")
+
+	return
+}
+
+func (self *InitialSyncer) syncBlocksInBatch(blocks []int64, ctx context.Context) (lastSyncedHeight int64) {
```

Comment:
you should divide this code into tasks. Each task doing one small thing. Otherwise it's hard to review.